File: bsp-testkit2/bsp-testkit/client/Session.scala
package org.jetbrains.bsp.testkit.client

import ch.epfl.scala.bsp4j.BspConnectionDetails
import com.google.gson.Gson
import org.eclipse.lsp4j.jsonrpc.Launcher
import org.jetbrains.bsp.testkit.client.Session.readBspConnectionDetails

import java.io.File
import java.nio.file.Path
import java.util.concurrent.{ExecutorService, Executors}
import scala.concurrent.{ExecutionContext, Future}
import scala.io.Source
import scala.jdk.FutureConverters.CompletionStageOps

/** A session is a "physical" connection to a BSP server. It must be closed when it is no longer
 * needed. The user is responsible for maintaining the correct BSP life-cycle.
 *
 * @param workspacePath
 * the path to the workspace
 */
class Session(val workspacePath: Path) extends AutoCloseable {
  private val workspaceFile = workspacePath.toFile
  private val connectionDetails = readBspConnectionDetails(workspaceFile)

  private val process =
    new ProcessBuilder(connectionDetails.getArgv)
      .directory(workspaceFile)
      .start() // side-effect: starts the process

  val client: MockClient = new MockClient()

  private val executor: ExecutorService = Executors.newCachedThreadPool()

  private val launcher = new Launcher.Builder[MockServer]()
    .setRemoteInterface(classOf[MockServer])
    .setExecutorService(executor)
    .setInput(process.getInputStream)
    .setOutput(process.getOutputStream)
    .setLocalService(client)
    .create()

  launcher.startListening() // side-effect: server starts listening for messages

  val server: MockServer = launcher.getRemoteProxy

  def serverClosed: Future[SessionResult] = process
    .onExit()
    .asScala
    .map(process =>
      SessionResult(process.exitValue(), Source.fromInputStream(process.getErrorStream).mkString)
    )(ExecutionContext.global)

  def close(): Unit = {
    executor.shutdown()
    process.getInputStream.close()
    process.getOutputStream.close()
    process.getErrorStream.close()
    process.destroy()
  }
}

case class SessionResult(exitCode: Int, stderr: String)

object Session {
  val bspVersion = "2.0.0"

  private val BspWorkspaceConfigDirName = ".bsp"
  private val BspWorkspaceConfigFileExtension = ".json"
  private val gson = new Gson()

  private def readBspConnectionDetails(workspace: File): BspConnectionDetails = {
    val bspDir = new File(workspace, BspWorkspaceConfigDirName)
    assert(bspDir.exists, s"$workspace is not a BSP workspace")
    assert(bspDir.isDirectory, s"$workspace is not a BSP workspace")

    val configFiles =
      bspDir.listFiles(file => file.getName.endsWith(BspWorkspaceConfigFileExtension))
    assert(
      configFiles.size == 1,
      s"$workspace should have exactly one BSP workspace configuration file"
    )

    val configFile = configFiles.head
    assert(
      configFile.isFile,
      s"$workspace should have exactly one BSP workspace configuration file"
    )
    assert(
      configFile.canRead,
      s"$workspace should have exactly one BSP workspace configuration file"
    )

    gson.fromJson(Source.fromFile(configFile).bufferedReader(), classOf[BspConnectionDetails])
  }
}

File: testkit/client/Session.kt
package org.jetbrains.bsp.testkit.client

import ch.epfl.scala.bsp4j.BspConnectionDetails
import ch.epfl.scala.bsp4j.BuildClient
import ch.epfl.scala.bsp4j.BuildServer
import com.google.gson.Gson
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.future.asDeferred
import org.eclipse.lsp4j.jsonrpc.Launcher
import java.io.File
import java.nio.file.Path
import java.util.concurrent.Executors


/**
 * A session is a "physical" connection to a BSP server. It must be closed when it is no longer
 * needed. The user is responsible for maintaining the correct BSP life-cycle.
 */
class Session<Server : BuildServer, Client : BuildClient>(
  val workspacePath: Path,
  val client: Client,
  serverClass: Class<Server>
) : AutoCloseable {
  private val workspaceFile = workspacePath.toFile()
  private val connectionDetails = readBspConnectionDetails(workspaceFile)

  private val logFile = File.createTempFile("bsp-testkit", ".log")

  val process: Process = ProcessBuilder(connectionDetails.argv)
    .directory(workspaceFile)
    .redirectError(logFile)
    .start()

  private val executor = Executors.newCachedThreadPool()
  private val launcher = Launcher.Builder<Server>()
    .setRemoteInterface(serverClass)
    .setExecutorService(executor)
    .setInput(process.inputStream)
    .setOutput(process.outputStream)
    .setLocalService(client)
    .create()

  init {
    launcher.startListening()
  }

  val server: Server = launcher.remoteProxy

  val serverClosed: Deferred<SessionResult> = process.onExit().thenApply {
    SessionResult(process.exitValue(), logFile.readText())
  }.asDeferred()

  override fun close() {
    executor.shutdown()
    process.inputStream.close()
    process.outputStream.close()
    process.errorStream.close()
    process.destroy()
  }

  companion object {
    private const val BspWorkspaceConfigDirName = ".bsp"
    private const val BspWorkspaceConfigFileExtension = ".json"
    private val gson = Gson()

    private fun readBspConnectionDetails(workspace: File): BspConnectionDetails {
      val bspDir = File(workspace, BspWorkspaceConfigDirName)
      require(bspDir.exists() && bspDir.isDirectory)

      val configFiles = bspDir.listFiles { _, name -> name.endsWith(BspWorkspaceConfigFileExtension) }
      require(configFiles != null)
      require(configFiles.size == 1)

      val configFile = configFiles.first()
      require(configFile.isFile && configFile.canRead())

      return gson.fromJson(configFile.bufferedReader(), BspConnectionDetails::class.java)
    }
  }
}

data class SessionResult(val exitCode: Int, val stderr: String)
